private/bufpkg/bufmodule/bufmoduleref/bufmoduleref.go

```diff
@@ -364,6 +383,54 @@ func ValidateModulePinsUniqueByIdentity(modulePins []ModulePin) error {
 	return nil
 }
 
+// ValidateModulePinsConsistentDigests verifies that module pins to the same commit don't change digests.
+// This is important to avoid MITM issues, where the module digest stored in a buf.lock file doesn't match
+// the module pin returned from the BSR.
+// Returns DigestChangedError if any valid digest changed from the buf.lock file for the same dependency commit.
+func ValidateModulePinsConsistentDigests(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+	modulePins []ModulePin,
+) error {
+	currentConfig, err := buflock.ReadConfig(ctx, bucket)
+	if err != nil {
+		if storage.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	if len(currentConfig.Dependencies) == 0 {
+		return nil
+	}
+	currentIdentityAndCommitToDigest := make(map[string]string, len(currentConfig.Dependencies))
+	for _, dep := range currentConfig.Dependencies {
+		// Ignore dependencies with no digest
+		if dep.Digest == "" {
+			continue
+		}
+		// Ignore dependencies with an invalid digest.
+		// We want to replace these with a valid digest.
+		if _, err := manifest.NewDigestFromString(dep.Digest); err != nil {
+			continue
+		}
+		key := fmt.Sprintf("%s/%s/%s:%s", dep.Remote, dep.Owner, dep.Repository, dep.Commit)
+		currentIdentityAndCommitToDigest[key] = dep.Digest
+	}
+	for _, pin := range modulePins {
+		if pin.Digest() == "" {
+			continue
+		}
+		key := fmt.Sprintf("%s:%s", pin.IdentityString(), pin.Commit())
+		if currentDigest, ok := currentIdentityAndCommitToDigest[key]; ok && currentDigest != pin.Digest() {
+			return &DigestChangedError{
```

Comment:
Nit: Would it be worth to group them in `multierr`? So all errors go out at once? Not sure how that would impact the caller ability to do `errors.As`